# circRNA_NB
Code supporting the analysis for Fuchs, et al., Defining the landscape of circular RNAs in neuroblastoma unveils a global suppressive function of MYCN. The code and scripts are written by Filippos Klironomos.

The following scripts are available:

 1_circRNA.R - Identifying and defining the circRNA sequences in the cell line and tumors including flanking introns

2_circular_linear_junction.R - Constructing the linear and circular junctions and estimating the gene expression 

3_DEanalysis.R - Differential expression analysis between genes and circRNAs 

4_RBP_motif.R - circRNA -RBP motif analysis

5_tetinductionmycn.R - Basic statistical analysis on the results of the tet-induction experiment

6_proliferative_index.Rdata - Calculation of the proliferative index based on a published expression signature



Please do not hesitate to contact the Schulte Lab: https://kinderonkologie.charite.de/en/research/schulte_group/
